x-pack/plugins/maps/public/classes/sources/table_source/table_source.ts

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import uuid from 'uuid';
+import { FIELD_ORIGIN, SOURCE_TYPES, VECTOR_SHAPE_TYPE } from '../../../../common/constants';
+import {
+  MapExtent,
+  MapFilters,
+  MapQuery,
+  TableSourceDescriptor,
+  VectorJoinSourceRequestMeta,
+  VectorSourceSyncMeta,
+} from '../../../../common/descriptor_types';
+import { Adapters } from '../../../../../../../src/plugins/inspector/common/adapters';
+import { ITermJoinSource } from '../term_join_source';
+import { BucketProperties, PropertiesMap } from '../../../../common/elasticsearch_util';
+import { IField } from '../../fields/field';
+import { Query } from '../../../../../../../src/plugins/data/common/query';
+import {
+  AbstractVectorSource,
+  BoundsFilters,
+  GeoJsonWithMeta,
+  IVectorSource,
+  SourceTooltipConfig,
+} from '../vector_source';
+import { DataRequest } from '../../util/data_request';
+import { InlineField } from '../../fields/inline_field';
+
+export class TableSource extends AbstractVectorSource implements ITermJoinSource, IVectorSource {
+  static type = SOURCE_TYPES.TABLE_SOURCE;
+
+  static createDescriptor(descriptor: Partial<TableSourceDescriptor>): TableSourceDescriptor {
+    return {
+      type: SOURCE_TYPES.TABLE_SOURCE,
+      __rows: descriptor.__rows || [],
+      __columns: descriptor.__columns || [],
+      term: descriptor.term || '',
+      id: descriptor.id || uuid(),
+    };
+  }
+
+  readonly _descriptor: TableSourceDescriptor;
+
+  constructor(descriptor: Partial<TableSourceDescriptor>, inspectorAdapters?: Adapters) {
+    const sourceDescriptor = TableSource.createDescriptor(descriptor);
+    super(sourceDescriptor, inspectorAdapters);
+    this._descriptor = sourceDescriptor;
+  }
+
+  async getDisplayName(): Promise<string> {
+    // no need to localize. this is never rendered.
+    return `table source ${uuid()}`;
+  }
+
+  getSyncMeta(): VectorSourceSyncMeta | null {
+    return null;
+  }
+
+  async getPropertiesMap(
+    searchFilters: VectorJoinSourceRequestMeta,
+    leftSourceName: string,
+    leftFieldName: string,
+    registerCancelCallback: (callback: () => void) => void
+  ): Promise<PropertiesMap> {
+    const propertiesMap: PropertiesMap = new Map<string, BucketProperties>();
+
+    const fieldNames = await this.getFieldNames();
+
+    for (let i = 0; i < this._descriptor.__rows.length; i++) {
+      const row: { [key: string]: string | number } = this._descriptor.__rows[i];
```

Comment:
Do you need to define row type like this? this._descriptor.__rows is already typed.

Reply:
will keep to improve readability, since it's a little intricate how keys/values get juggled around.